File: src/main/GO/leetcode/lists/hot200/LeetCode333.go
package hot200

import "math"

/**
 * @author wujingxinit@outlook.com
 * @date 2023/3/16 17:13
 */

func largestBSTSubtree(root *TreeNode) (ans int) {
	var dfs func(node *TreeNode) (bool, int, int, int) // 是否搜索树 总节点数量 最小节点值 最大节点值
	dfs = func(node *TreeNode) (bool, int, int, int) {
		if node == nil {
			return true, 0, math.MaxInt, math.MinInt
		}
		left, leftCnt, leftMin, leftMax := dfs(node.Left)
		right, rightCnt, rightMin, rightMax := dfs(node.Right)
		cur := left && right && node.Val > leftMax && node.Val < rightMin
		curCnt := leftCnt + rightCnt + 1
		// curMin curMax 的作用？
		curMin := Min(node.Val, Min(leftMin, rightMin))
		curMax := Max(node.Val, Max(leftMax, rightMax))
		if cur {
			ans = Max(ans, curCnt)
		}
		return cur, curCnt, curMin, curMax
	}
	dfs(root)
	return
}

File: src/main/java/leetcode/lists/hot200/LeetCode333.java
package leetcode.lists.hot200;

import common.TreeNode;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/3/16 16:09
 */
public class LeetCode333 {


    public int largestBSTSubtree(TreeNode root) {
        if (root == null) {
            return 0;
        }
        if (dfs(root, Integer.MIN_VALUE, Integer.MAX_VALUE)) {
            cnt(root);
        }
        return Math.max(largestBSTSubtree(root.left), largestBSTSubtree(root.right));
    }

    private int cnt(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return cnt(root.left) + cnt(root.right) + 1;
    }


    public boolean dfs(TreeNode node, Integer min, Integer max) {
        if (node == null) {
            return true;
        }
        if (node.val <= min || node.val >= max) {
            return false;
        }
        return dfs(node.left, min, node.val) && dfs(node.right, node.val, max);
    }

    //=====opt====
    int res = 0;

    public int largestBSTSubtree_opt(TreeNode root) {
        dfs(root);
        return res;
    }

    //以当前节点为根是否是BST
    int[] dfs(TreeNode root) {
        //空节点也是一颗BST
        if (root == null) {
            return new int[]{1, Integer.MAX_VALUE, Integer.MIN_VALUE, 0};
        }
        int[] left = dfs(root.left);
        int[] right = dfs(root.right);
        // [0]是否BST
        // [1]当前为根的BST最小值
        // [2]当前为根的BST最大值
        // [3]该BST节点总数
        // 左右子树是BST，并且val满足大于左最大值，小于右最小值
        int[] cur = new int[4];
        if (left[0] == 1 && right[0] == 1 && root.val > left[2] && root.val < right[1]) {
            cur[0] = 1;//置1 true
            cur[1] = Math.min(left[1], root.val);//更新最小值
            cur[2] = Math.max(right[2], root.val);//更新最大值
            cur[3] = left[3] + right[3] + 1;//更新节点数量
            res = Math.max(cur[3], res);//更新全局变量
        }
        return cur;
    }
}
